packages/dds/shared-object-base/src/sharedObject.ts

```diff
@@ -208,30 +208,56 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
      * {@inheritDoc (ISharedObject:interface).summarize}
      */
     public summarize(fullTree: boolean = false, trackState: boolean = false): IChannelSummarizeResult {
+        // Set _isSummarizing to true. This flag is used to ensure that we only use SummarySerializer (created below)
+        // to serialize handles in this object's data. The routes of these serialized handles are outbound routes
+        // to other Fluid objects.
+        assert(!this._isSummarizing, "Possible re-entrancy! Summary should not already be in progress.");
         this._isSummarizing = true;
 
-        /**
-         * Create a SummarySerializer that will be used to serialize IFluidHandles in this object. SummarySerializer
-         * tracks the routes of all handles that it serializes. These represent routes to referenced Fluid object.
-         */
         const serializer = new SummarySerializer(this.runtime.channelsRoutingContext);
         const snapshot: ITree = this.snapshotCore(serializer);
         const summaryTree = convertToSummaryTreeWithStats(snapshot, fullTree);
 
-        /**
-         * We need to add this channel's garbage collection node to the summarize result.
-         * The outbound routes of this channel are all the routes of all the handles that are tracked by the
-         * SummarySerializer created above.
-         */
-        const gcNodes: IGraphNode[] = [
-            { id: "/", outboundRoutes: serializer.getSerializedRoutes() },
-        ];
+        // Add this channel's garbage collection data to the summarize result. The outbound routes of this channel are
+        // all the routes of all the handles that are tracked by the SummarySerializer above.
+        const gcData: IGCData = {
+            gcNodes: { "/": serializer.getSerializedRoutes() },
+        };
 
+        assert(this._isSummarizing, "Possible re-entrancy! Summary should have been in progress.");
         this._isSummarizing = false;
```

Comment:
this should go in a finally, otherwise if an exception is thrown the object will always have this._isSummarizing = true;